### PR TITLE
IQSS/9185 contact email updates

### LINKED ISSUE #9185: Feature Request/Idea: Improve Contact Form Email
**Overview of the Feature Request** The current contact form mechanism has some limitations that QDR and others are interested in removing. These include:

- Each contact is sent a separate email, so none of them know who else received a request and there is no single group email where a 'reply all' can be done
- When contacts for a collection or dataset exists, the repository does not receive (via to or cc) an email. For curated repositories, this means administrators are not aware when requests come in
- When the form emails the repository, it uses the :SystemEmail address. [For some repositories](https://groups.google.com/g/dataverse-community/c/b9y7lg2J9-w/m/SrHEk3hFAgAJ), there is interest in making the SystemEmail a no-reply address which would not be useful for support emails
- In determining the form of the name in a greeting, the current code assumes that any comma in the name string implies a person with a \<last name\>,\<first name\> order and it reverses the two parts. This is not correct for names such as 'IBM, Inc.'
- while there is a Feedback API call, it doesn't work and isn't documented.

QDR will submit a PR with changes addressing these issues. As usual the PR can be adjusted/split as desired.

### PR DESCRIPTION
**Pull request by qqmyers**
NOTE: per review, the settings have been replaced with Microprofile entries. The documentation has been updated but the PR notes below have not been updated!

**What this PR does / why we need it**: This PR includes several enhancements to the emails from the contact forms that have been implemented and deployed at QDR.  It also include fixes to make the /admin/feedback endpoint work.

Specific changes:
- new :CCSupportOnContactEmails setting which will include the support email address as a CC: entry when contact/feedback emails are sent to the contacts for a collection, dataset, or datafile. The setting also adds a CC line in the form so that users are aware that it will happen
- new :SupportEmail setting that allows a separate email, distinct from the :SystemEmail to be used as the from address in emails from the contact form/ feedback api. (Useful when admins want :SystemEmail to be a no-reply address. and/or to direct contact form emails to a different set of people/different support system.)
- Switch to sending one email to all contacts instead of sending an email to each. This is hopefully helpful in general (contacts can see who else is being informed) and simplifies adding a cc to support.
- Use of the PersonOrOrg algorithm to decide if a name refers to a person, in which case a \<lastname\>,\<firstname\> order should be reversed in the greeting ("Hello Homer Simpson"), or an Org (where a name with commas should not be changed).
- Addition of an algorithm to separate names with commas and include a final "and" when there are multiple contacts (e.g. "Hello Homer Simpson, Marge Simpson, and Bart Simpson")
- Update of tests to match
- Bug fixes/completion of the /admin/feedback endpoint so it sends emails.
- Docs for the api call and new settings, release notes.

**Which issue(s) this PR closes**:

Closes #9185

**Special notes for your reviewer**:

Depends on #9089 for the PersonOrOrgUtil changes. I've copied that class here so it will build.
Any of this can be broken out/dropped if the whole set of changes is not desired.

**Suggestions on how to test this**:
- Try the contact form and API call. Use multiple contacts for some tests to verify that one email is sent. 
- Try the CC setting and confirm there's a cc in the UI and the email. 
- Try the :SupportEmail setting and confirm that emails from the contact forms (and not other emails) are from that address.
- Try Org names like "IBM, Inc." to make sure they don't get flipped to "Hello Inc. IBM" in the greeting line in the emails.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

 Using the CCSupportOnContactEmails setting adds a line to the contact form: 
image

**Is there a release notes update needed for this change?**:
included.

**Additional documentation**:

### CODE REVIEW

**Comment by mreekie** [comment]
Grooming:

This issue was in the New or No Status columns on the Dataverse Global Backlog Board.
Those columns have been removed.
This issue has been removed from the board

This is NOT a reflection of the priority of these issues.
However, as we have worked on the grooming process, it became clear that these issues which are in columns that are not stewarded will likely not get prioritized or sized.

To get this item onto the Dataverse Global Backlog Board, please reach out to one of the Stewards on the board.
You can also reach out to @mreekie and I can connect you with the steward of the appropriate queue

**Comment by jggautier** [comment]
> new :SupportEmail setting that allows a separate email, distinct from the :SystemEmail to be used as the from address in emails from the contact form/ feedback api. (Useful when admins want :SystemEmail to be a no-reply address. and/or to direct contact form emails to a different set of people/different support system.)

- I've been excited about this issue because of this line! Right now, the same email address is used when:
Dataverse sends automated email notifications to users when certain things happen ("your account was created", "you were given a role", etc.), and
- When people click the Support link at the top of the page to email the installation admins

So with this new :SupportEmail setting, an installation can use two different email addresses, one for the automated email notifications and another for when people email "support". That would help address an issue at the Harvard repository where I think one reason why email clients keep marking email sent from our one email address (support[at]dataverse.harvard.edu) as spam is because it notices that people aren't opening the automated email notifications that the installation sends, and this also means that emails that installation staff send to users using that same email address get marked as spam. This affects repository staff's ability to contact users. This is one of the many potential reasons why email clients might keep marking emails from our support email address as spam. I'm talking with repository staff about the other reasons and possible solutions.

Basically, using two different emails (one email address for the automated notifications and another for when depositors and repository staff contact each other) might help increase the odds that depositors see the emails that repository staff send.

**Comment by qqmyers** [comment]
Relatively simple changes to how the contact form sends emails (see changes in the description). Sizing at 10 to allow review (including the code making the feedback admin API call functional and changing to send one email to all contacts) and testing of the two new settings.

File: doc/sphinx-guides/source/api/native-api.rst [hunk 1]
```diff
@@ -4497,3 +4497,26 @@ A curl example using allowing access to a dataset's metadata
 
 Please see :ref:`dataverse.api.signature-secret` for the configuration option to add a shared secret, enabling extra
 security.
+
+Send Feedback To Contact(s)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (pdurbin) [inline doc/sphinx-guides/source/api/native-api.rst:4503]: We should probabl add something like this:

See also Return a Dataset to Author

It's very similar: https://guides.dataverse.org/en/5.13/api/native-api.html#return-a-dataset-to-author

And we should probably add a see also from there back to here.

File: src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java [hunk 2]
```diff
@@ -174,12 +174,7 @@ public enum Key {
          *
          */
         SearchRespectPermissionRoot,
-        /**
-         * Solr hostname and port, such as "localhost:8983".
-         * @deprecated New installations should not use this database setting, but use {@link JvmSettings#SOLR_HOST}
-         *             and {@link JvmSettings#SOLR_PORT}.
-         */
-        @Deprecated(forRemoval = true, since = "2022-12-23")
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java:182]: Why was this deprecation notice removed? I assume it's there for a reason.

File: src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java [hunk 3]
```diff
@@ -568,15 +563,26 @@ Whether Harvesting (OAI) service is enabled
...
+         * Optional email address to use with the support form/ feedback api emails. Default is to use the SystemEmail account.
+         * 
+         */
+        
+        SupportEmail,
+        /*
+         * Whether the repository support team should be CC'd on requests created in the
+         * contact form, when email is sent TO the listed Dataverse, Dataset, or DataFile contacts. Default is false.
+         */
+        CCSupportOnContactEmails
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java:585]: Instead of adding new database settings, please make these MPCONFIG settings.

**Review by pdurbin — CHANGES_REQUESTED** [review]
I was about to test this but then I noticed that new database settings are being added. Please change these to MPCONFIG.

I also left a bit more feedback on docs.

Overall, these seem like nice features! And hooray for tests! 🎉

File: src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java (hunk 2 above)
Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java]: merge error

**Comment by pdurbin** [comment]
To run a dev email server (MailDev) locally, this should work:

```
$ docker-compose -f docker-compose-dev.yml up dev_smtp
Creating network "dataverse_dataverse" with driver "bridge"
Recreating dev_smtp ... done
Attaching to dev_smtp
dev_smtp                | MailDev using directory /mail
dev_smtp                | MailDev webapp running at http://0.0.0.0:1080
dev_smtp                | MailDev SMTP Server running at 0.0.0.0:25
```

The UI is at http://localhost:1080 and looks like this:

Screen Shot 2023-04-28 at 10 20 15 AM

File: src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java [hunk 4]
```diff
@@ -111,7 +111,7 @@ public void sendMail(String host, String reply, String to, String subject, Strin
             me.printStackTrace(System.out);
         }
     }
-
+*/
     @Resource(name = "mail/notifyMailSession")
```
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java:115]: Any reason we need to keep this fully commented out method around? (Also down around line 180)

File: src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java [hunk 5]
```diff
@@ -177,11 +177,11 @@ public InternetAddress getSystemAddress() {
...
-    public void sendMail(String from, String to, String subject, String messageText) {
-        sendMail(from, to, subject, messageText, new HashMap<>());
+/*    public void sendMail(String from, String to, String cc, String subject, String messageText) {
+        sendMail(from, to, cc, subject, messageText, new HashMap<>());
     }
-
-    public void sendMail(String reply, String to, String subject, String messageText, Map<Object, Object> extraHeaders) {
+*/
+    public void sendMail(String reply, String to, String cc, String subject, String messageText, Map<Object, Object> extraHeaders) {
         try {
```
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java:185]: The only two usages I see for this method pass null for the extra headers param. Is it needed?

Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java:185]: OK. It looks like it's been around forever. hmm

**Review by sekmiller — CHANGES_REQUESTED** [review]
Looks good. thanks for putting in the mp config. Just a couple more comments on some unused code.

File: src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java (hunk 5 above)
Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java:150]: Removed commented-out code and remove the extra param and code here.

**Comment by qqmyers** [comment]
FWIW: It looks like the failing unit test is because https://ddialliance.org/Specification/DDI-Codebook/2.5/XMLSchema/codebook.xsd couldn't be read - as has been seen on other branches lately.

**Review by pdurbin — APPROVED** [review]
Just hitting "Approve" to clear my "requested changes" status. But then I'll put this back where it was (In Review).

**Review by sekmiller — APPROVED** [review]

**Comment by kcondon** [comment]
Issues found:
1. Feedback to contacts API example shows demo as server_url but endpoint only supports local host 
3. Feedback to contacts API example has jsonid on the command line example in front of $SERVER_URL, no space. Should not be there as it is in the json. 
4. 3. Calling the Feedback to contacts API endpoint results in null ptr in server.log
[emailapi_err.txt](https://github.com/IQSS/dataverse/files/11401196/emailapi_err.txt)
